File: maaden_strategy.py
"""
Daily trading algorithm for Maaden (Saudi Arabian Mining Co., Tadawul: 1211.SR).

Strategy: SMA crossover for trend direction, filtered by RSI to avoid buying
into blow-off-top conditions, with a trailing stop for risk management
(cuts losers early, lets winners run instead of capping them with a fixed
take-profit). Long-only (matches Tadawul, which has no retail short-selling).

WHY THESE RULES (full search history, for honesty about what was tried):
Roughly 1,300 parameter combinations were tested across FOUR different rule
families, each evaluated across three separate time windows (2018-2026,
2025-2026, 2024-2026) so results couldn't be cherry-picked from one lucky
period:
  1. Fixed stop-loss (4%) + fixed take-profit (8%) + RSI exit on overbought
     - net LOSS over 2024-2026 (-12.3%). Take-profit capped winners early;
       RSI-overbought exit kicked out of strong trends prematurely.
  2. SMA crossover + RSI filter + 8% TRAILING stop (no fixed take-profit,
     no RSI exit) - THIS VERSION. Profitable in every window tested:
       2018-2026: +161.2%  |  2025-2026: +21.6%  |  2024-2026: +17.5%
       (buy & hold for comparison: +258.4% | +22.7% | +23.7%)
  3. Same as #2 plus a long-term (100/150/200-day SMA) regime filter to
     avoid trading against the major trend - WORSE than #2 in the recent
     windows (delayed entries missed early moves): +7.6% / +14.9% best case.
  4. "Stay invested by default, only exit on a confirmed severe breakdown"
     (200-day SMA break confirmed over N days, or a large drawdown from
     peak) - worst-case return across the three windows was only +0.7%,
     because Maaden's normal pullbacks have run 25-40% deep within its own
     uptrend, so almost any "crash" threshold tight enough to react quickly
     also fires on ordinary noise.

Version #2 (below) is the one that survived: it is the only rule set that
stayed solidly profitable in all three windows without relying on one
period's luck. It still does not beat plain buy-and-hold on raw return in
any window - Maaden's trend has been clean enough that just holding wins
on that single metric. What #2 offers instead is a smaller, real edge:
consistent profitability with fewer, higher-conviction trades, and it
never turned an actual loss the way approach #1 did. A backtest describes
the past; it is not a promise about the future.

Usage:
    python maaden_strategy.py
"""

import numpy as np
import pandas as pd
import yfinance as yf

TICKER = "1211.SR"
START = "2018-01-01"

FAST_SMA = 10
SLOW_SMA = 30
RSI_PERIOD = 14
RSI_MAX_ENTRY = 85       # don't enter if RSI already this hot (blow-off top)

TRAILING_STOP_PCT = 0.08  # exit if price falls 8% from its peak since entry

INITIAL_CAPITAL = 100_000
COMMISSION_PCT = 0.001  # 0.1% per trade (round-trip fee approximation)


def fetch_data(ticker: str, start: str) -> pd.DataFrame:
    df = yf.download(ticker, start=start, progress=False, auto_adjust=True)
    if isinstance(df.columns, pd.MultiIndex):
        df.columns = df.columns.get_level_values(0)
    df = df.dropna(subset=["Close"])
    return df


def rsi(series: pd.Series, period: int) -> pd.Series:
    delta = series.diff()
    gain = delta.clip(lower=0)
    loss = -delta.clip(upper=0)
    avg_gain = gain.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()
    avg_loss = loss.ewm(alpha=1 / period, min_periods=period, adjust=False).mean()
    rs = avg_gain / avg_loss
    return 100 - (100 / (1 + rs))


def add_indicators(df: pd.DataFrame) -> pd.DataFrame:
    df["sma_fast"] = df["Close"].rolling(FAST_SMA).mean()
    df["sma_slow"] = df["Close"].rolling(SLOW_SMA).mean()
    df["rsi"] = rsi(df["Close"], RSI_PERIOD)
    return df


def generate_signals(df: pd.DataFrame) -> pd.DataFrame:
    """
    Entry (go long): fast SMA crosses above slow SMA AND RSI isn't already
    at a blow-off-top extreme.
    Exit: fast SMA crosses below slow SMA (trend reversal), OR the trailing
    stop is hit (handled in backtest loop).
    """
    trend_up = df["sma_fast"] > df["sma_slow"]
    cross_up = trend_up & ~trend_up.shift(1).fillna(False)
    cross_down = ~trend_up & trend_up.shift(1).fillna(False)

    df["entry_signal"] = cross_up & (df["rsi"] < RSI_MAX_ENTRY)
    df["exit_signal"] = cross_down
    return df


def backtest(df: pd.DataFrame) -> dict:
    cash = INITIAL_CAPITAL
    shares = 0
    entry_price = 0.0
    peak_price = 0.0
    equity_curve = []
    trades = []

    for date, row in df.iterrows():
        price = row["Close"]

        if shares > 0:
            peak_price = max(peak_price, price)
            change = (price - entry_price) / entry_price
            trail_hit = price <= peak_price * (1 - TRAILING_STOP_PCT)

            if trail_hit or row["exit_signal"]:
                proceeds = shares * price * (1 - COMMISSION_PCT)
                cash += proceeds
                trades.append({
                    "exit_date": date,
                    "exit_price": price,
                    "return_pct": change * 100,
                    "reason": "trailing_stop" if trail_hit else "signal",
                })
                shares = 0

        elif shares == 0 and row.get("entry_signal", False):
            shares = int((cash * (1 - COMMISSION_PCT)) // price)
            if shares > 0:
                cost = shares * price * (1 + COMMISSION_PCT)
                cash -= cost
                entry_price = price
                peak_price = price
                trades.append({"entry_date": date, "entry_price": price, "shares": shares})

        equity_curve.append(cash + shares * price)

    df["equity"] = equity_curve
    final_value = df["equity"].iloc[-1]

    completed = [t for t in trades if "return_pct" in t]
    wins = [t for t in completed if t["return_pct"] > 0]

    returns = df["equity"].pct_change().dropna()
    sharpe = (returns.mean() / returns.std()) * np.sqrt(252) if returns.std() > 0 else 0

    running_max = df["equity"].cummax()
    drawdown = (df["equity"] - running_max) / running_max
    max_drawdown = drawdown.min()

    years = (df.index[-1] - df.index[0]).days / 365.25
    cagr = (final_value / INITIAL_CAPITAL) ** (1 / years) - 1 if years > 0 else 0

    return {
        "final_value": final_value,
        "total_return_pct": (final_value / INITIAL_CAPITAL - 1) * 100,
        "cagr_pct": cagr * 100,
        "sharpe_ratio": sharpe,
        "max_drawdown_pct": max_drawdown * 100,
        "num_trades": len(completed),
        "win_rate_pct": (len(wins) / len(completed) * 100) if completed else 0,
        "avg_trade_return_pct": np.mean([t["return_pct"] for t in completed]) if completed else 0,
        "trades": trades,
    }


def buy_and_hold(df: pd.DataFrame) -> float:
    shares = INITIAL_CAPITAL // df["Close"].iloc[0]
    leftover = INITIAL_CAPITAL - shares * df["Close"].iloc[0]
    return shares * df["Close"].iloc[-1] + leftover


def _monthly_contribution_dates(df: pd.DataFrame) -> set:
    """First trading day of each calendar month in the data."""
    return set(df.groupby(df.index.to_period("M")).apply(lambda g: g.index[0]))


def dca_backtest(df: pd.DataFrame, monthly_amount: float) -> dict:
    """
    Same entry/exit rules as backtest(), but capital arrives as a fixed
    contribution on the first trading day of each month instead of all
    at once. New cash only gets deployed on the next entry signal - while
    a position is already open, contributions just wait in cash.
    """
    contribution_dates = _monthly_contribution_dates(df)

    cash = 0.0
    shares = 0
    entry_price = 0.0
    peak_price = 0.0
    total_invested = 0.0
    equity_curve = []
    invested_curve = []
    trades = []

    for date, row in df.iterrows():
        price = row["Close"]

        if date in contribution_dates:
            cash += monthly_amount
            total_invested += monthly_amount

        if shares > 0:
            peak_price = max(peak_price, price)
            change = (price - entry_price) / entry_price
            trail_hit = price <= peak_price * (1 - TRAILING_STOP_PCT)

            if trail_hit or row["exit_signal"]:
                cash += shares * price * (1 - COMMISSION_PCT)
                trades.append({
                    "exit_date": date,
                    "exit_price": price,
                    "return_pct": change * 100,
                    "reason": "trailing_stop" if trail_hit else "signal",
                })
                shares = 0

        elif shares == 0 and row.get("entry_signal", False):
            new_shares = int((cash * (1 - COMMISSION_PCT)) // price)
            if new_shares > 0:
                cash -= new_shares * price * (1 + COMMISSION_PCT)
                shares = new_shares
                entry_price = price
                peak_price = price
                trades.append({"entry_date": date, "entry_price": price, "shares": shares})

        equity_curve.append(cash + shares * price)
        invested_curve.append(total_invested)

    df["equity"] = equity_curve
    final_value = df["equity"].iloc[-1]
    completed = [t for t in trades if "return_pct" in t]
    wins = [t for t in completed if t["return_pct"] > 0]

    return {
        "total_invested": total_invested,
        "final_value": final_value,
        "profit": final_value - total_invested,
        "return_pct": (final_value / total_invested - 1) * 100 if total_invested else 0,
        "num_trades": len(completed),
        "win_rate_pct": (len(wins) / len(completed) * 100) if completed else 0,
    }


def dca_buy_and_hold(df: pd.DataFrame, monthly_amount: float) -> dict:
    """Every month, buy as many shares as the accumulated cash allows."""
    contribution_dates = _monthly_contribution_dates(df)

    cash = 0.0
    shares = 0
    total_invested = 0.0

    for date, row in df.iterrows():
        price = row["Close"]
        if date in contribution_dates:
            cash += monthly_amount
            total_invested += monthly_amount
            new_shares = int(cash // price)
            if new_shares > 0:
                cash -= new_shares * price
                shares += new_shares

    final_value = shares * df["Close"].iloc[-1] + cash
    return {
        "total_invested": total_invested,
        "final_value": final_value,
        "profit": final_value - total_invested,
        "return_pct": (final_value / total_invested - 1) * 100 if total_invested else 0,
        "shares": shares,
    }


def dca_scenario(ticker: str, start: str, monthly_amount: float):
    df = fetch_data(ticker, start)
    df = add_indicators(df)
    df = generate_signals(df)

    strat = dca_backtest(df, monthly_amount)
    bh = dca_buy_and_hold(df, monthly_amount)
    months = len(_monthly_contribution_dates(df))

    print(f"\n=== Monthly DCA of {monthly_amount:.0f} SAR into {ticker} ===")
    print(f"Period: {df.index[0].date()} to {df.index[-1].date()} ({months} contributions)")
    print(f"Total invested: {strat['total_invested']:,.2f} SAR")
    print()
    print("--- Strategy (SMA crossover + RSI + trailing stop) ---")
    print(f"Final value: {strat['final_value']:,.2f} SAR")
    print(f"Profit:      {strat['profit']:,.2f} SAR ({strat['return_pct']:.2f}%)")
    print(f"Trades: {strat['num_trades']}, Win rate: {strat['win_rate_pct']:.1f}%")
    print()
    print("--- Buy & hold (invest every contribution immediately) ---")
    print(f"Final value: {bh['final_value']:,.2f} SAR")
    print(f"Profit:      {bh['profit']:,.2f} SAR ({bh['return_pct']:.2f}%)")
    print(f"Shares held: {bh['shares']}")
    return strat, bh


def main():
    df = fetch_data(TICKER, START)
    df = add_indicators(df)
    df = generate_signals(df)
    result = backtest(df)
    bh_final = buy_and_hold(df)

    print(f"\n=== Maaden ({TICKER}) SMA{FAST_SMA}/{SLOW_SMA} + RSI Strategy Backtest ===")
    print(f"Period: {df.index[0].date()} to {df.index[-1].date()}")
    print(f"Initial capital:        {INITIAL_CAPITAL:,.0f} SAR")
    print(f"Final strategy value:   {result['final_value']:,.0f} SAR")
    print(f"Buy & hold final value: {bh_final:,.0f} SAR")
    print(f"Total return:           {result['total_return_pct']:.2f}%")
    print(f"CAGR:                   {result['cagr_pct']:.2f}%")
    print(f"Sharpe ratio:           {result['sharpe_ratio']:.2f}")
    print(f"Max drawdown:           {result['max_drawdown_pct']:.2f}%")
    print(f"Number of trades:       {result['num_trades']}")
    print(f"Win rate:               {result['win_rate_pct']:.1f}%")
    print(f"Avg trade return:       {result['avg_trade_return_pct']:.2f}%")

    last = df.iloc[-1]
    print(f"\n=== Latest snapshot ({df.index[-1].date()}) ===")
    print(f"Close: {last['Close']:.2f} | SMA{FAST_SMA}: {last['sma_fast']:.2f} | "
          f"SMA{SLOW_SMA}: {last['sma_slow']:.2f} | RSI: {last['rsi']:.1f}")
    if last["sma_fast"] > last["sma_slow"]:
        stance = "IN UPTREND (would hold/buy if RSI < %d)" % RSI_MAX_ENTRY
    else:
        stance = "IN DOWNTREND (stay out / would exit)"
    print(f"Current stance: {stance}")


if __name__ == "__main__":
    main()

File: maaden_strategy.cpp
// Daily trading algorithm for Maaden (1211.SR) - C++ port of maaden_strategy.py
//
// Same rules as the Python version: SMA10/30 crossover entry (filtered by
// RSI14 < 85), 8% trailing stop exit. This program runs the monthly-DCA
// scenario: 990 SAR contributed on the first trading day of every month,
// starting 2024-01-01, and compares the strategy against simply buying
// shares with every contribution (buy & hold).
//
// Price data is read from data_1211SR.csv (Date,Close), exported from the
// same Yahoo Finance source the Python version uses - C++ has no built-in
// way to fetch that data itself.
//
// Build:  g++ -O2 -std=c++17 -o maaden_strategy maaden_strategy.cpp
// Run:    ./maaden_strategy data_1211SR.csv 990

#include <algorithm>
#include <cmath>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <optional>
#include <sstream>
#include <string>
#include <vector>

constexpr int FAST_SMA = 10;
constexpr int SLOW_SMA = 30;
constexpr int RSI_PERIOD = 14;
constexpr double RSI_MAX_ENTRY = 85.0;
constexpr double TRAILING_STOP_PCT = 0.08;
constexpr double COMMISSION_PCT = 0.001;

struct Row {
    std::string date;
    double close;
    double sma_fast = std::nan("");
    double sma_slow = std::nan("");
    double rsi = std::nan("");
    bool trend_up = false;
    bool entry_signal = false;
    bool exit_signal = false;
    bool is_contribution = false;
};

struct Trade {
    std::string entry_date, exit_date;
    double entry_price = 0, exit_price = 0;
    double return_pct = 0;
    bool closed = false;
};

std::vector<Row> load_csv(const std::string& path) {
    std::ifstream f(path);
    if (!f) throw std::runtime_error("cannot open " + path);
    std::string line;
    std::getline(f, line);  // header
    std::vector<Row> rows;
    while (std::getline(f, line)) {
        if (line.empty()) continue;
        auto comma = line.find(',');
        Row r;
        r.date = line.substr(0, comma);
        r.close = std::stod(line.substr(comma + 1));
        rows.push_back(r);
    }
    return rows;
}

void compute_indicators(std::vector<Row>& rows) {
    int n = static_cast<int>(rows.size());

    // Simple moving averages
    for (int i = 0; i < n; ++i) {
        if (i >= FAST_SMA - 1) {
            double sum = 0;
            for (int k = i - FAST_SMA + 1; k <= i; ++k) sum += rows[k].close;
            rows[i].sma_fast = sum / FAST_SMA;
        }
        if (i >= SLOW_SMA - 1) {
            double sum = 0;
            for (int k = i - SLOW_SMA + 1; k <= i; ++k) sum += rows[k].close;
            rows[i].sma_slow = sum / SLOW_SMA;
        }
    }

    // RSI14, matching pandas .ewm(alpha=1/period, adjust=False).mean() semantics:
    // the recursion starts at the first real price delta (index 1), and the
    // first RSI_PERIOD deltas are needed before a value is reported (min_periods).
    double alpha = 1.0 / RSI_PERIOD;
    double avg_gain = 0, avg_loss = 0;
    int valid_count = 0;
    for (int i = 1; i < n; ++i) {
        double delta = rows[i].close - rows[i - 1].close;
        double gain = std::max(delta, 0.0);
        double loss = std::max(-delta, 0.0);
        if (valid_count == 0) {
            avg_gain = gain;
            avg_loss = loss;
        } else {
            avg_gain = avg_gain * (1 - alpha) + gain * alpha;
            avg_loss = avg_loss * (1 - alpha) + loss * alpha;
        }
        valid_count++;
        if (valid_count >= RSI_PERIOD) {
            double rs = avg_loss == 0 ? std::numeric_limits<double>::infinity() : avg_gain / avg_loss;
            rows[i].rsi = 100.0 - 100.0 / (1.0 + rs);
        }
    }

    // Trend / signals - undefined SMA comparisons are treated as "false",
    // matching how pandas evaluates comparisons against NaN.
    bool prev_trend_up = false;
    for (int i = 0; i < n; ++i) {
        bool has_smas = !std::isnan(rows[i].sma_fast) && !std::isnan(rows[i].sma_slow);
        bool trend_up = has_smas && (rows[i].sma_fast > rows[i].sma_slow);
        bool cross_up = trend_up && !prev_trend_up;
        bool cross_down = !trend_up && prev_trend_up;

        bool rsi_ok = !std::isnan(rows[i].rsi) && rows[i].rsi < RSI_MAX_ENTRY;
        rows[i].entry_signal = cross_up && rsi_ok;
        rows[i].exit_signal = cross_down;
        rows[i].trend_up = trend_up;

        prev_trend_up = trend_up;
    }

    // Mark first trading day of each calendar month as a contribution date.
    std::string prev_month;
    for (auto& r : rows) {
        std::string month = r.date.substr(0, 7);  // "YYYY-MM"
        if (month != prev_month) {
            r.is_contribution = true;
            prev_month = month;
        }
    }
}

struct DcaResult {
    double total_invested = 0;
    double final_value = 0;
    double profit = 0;
    double return_pct = 0;
    int num_trades = 0;
    double win_rate_pct = 0;
    std::vector<Trade> trades;
};

DcaResult dca_backtest(const std::vector<Row>& rows, double monthly_amount) {
    double cash = 0, shares = 0, entry_price = 0, peak_price = 0, total_invested = 0;
    std::vector<Trade> trades;

    for (const auto& row : rows) {
        double price = row.close;

        if (row.is_contribution) {
            cash += monthly_amount;
            total_invested += monthly_amount;
        }

        if (shares > 0) {
            peak_price = std::max(peak_price, price);
            double change = (price - entry_price) / entry_price;
            bool trail_hit = price <= peak_price * (1 - TRAILING_STOP_PCT);

            if (trail_hit || row.exit_signal) {
                cash += shares * price * (1 - COMMISSION_PCT);
                Trade t;
                t.exit_date = row.date;
                t.exit_price = price;
                t.return_pct = change * 100;
                t.closed = true;
                trades.push_back(t);
                shares = 0;
            }
        } else if (row.entry_signal) {
            double new_shares = std::floor((cash * (1 - COMMISSION_PCT)) / price);
            if (new_shares > 0) {
                cash -= new_shares * price * (1 + COMMISSION_PCT);
                shares = new_shares;
                entry_price = price;
                peak_price = price;
            }
        }
    }

    double final_value = cash + shares * rows.back().close;

    int completed = 0, wins = 0;
    for (const auto& t : trades) {
        if (t.closed) {
            completed++;
            if (t.return_pct > 0) wins++;
        }
    }

    DcaResult res;
    res.total_invested = total_invested;
    res.final_value = final_value;
    res.profit = final_value - total_invested;
    res.return_pct = total_invested > 0 ? (final_value / total_invested - 1) * 100 : 0;
    res.num_trades = completed;
    res.win_rate_pct = completed > 0 ? (100.0 * wins / completed) : 0;
    res.trades = trades;
    return res;
}

struct BuyHoldResult {
    double total_invested = 0;
    double final_value = 0;
    double profit = 0;
    double return_pct = 0;
    double shares = 0;
};

BuyHoldResult dca_buy_and_hold(const std::vector<Row>& rows, double monthly_amount) {
    double cash = 0, shares = 0, total_invested = 0;
    for (const auto& row : rows) {
        double price = row.close;
        if (row.is_contribution) {
            cash += monthly_amount;
            total_invested += monthly_amount;
            double new_shares = std::floor(cash / price);
            if (new_shares > 0) {
                cash -= new_shares * price;
                shares += new_shares;
            }
        }
    }
    double final_value = shares * rows.back().close + cash;
    BuyHoldResult res;
    res.total_invested = total_invested;
    res.final_value = final_value;
    res.profit = final_value - total_invested;
    res.return_pct = total_invested > 0 ? (final_value / total_invested - 1) * 100 : 0;
    res.shares = shares;
    return res;
}

int main(int argc, char** argv) {
    std::string csv_path = argc > 1 ? argv[1] : "data_1211SR.csv";
    double monthly_amount = argc > 2 ? std::stod(argv[2]) : 990.0;

    std::vector<Row> rows = load_csv(csv_path);
    compute_indicators(rows);

    int months = 0;
    for (const auto& r : rows) if (r.is_contribution) months++;

    DcaResult strat = dca_backtest(rows, monthly_amount);
    BuyHoldResult bh = dca_buy_and_hold(rows, monthly_amount);

    std::cout << std::fixed << std::setprecision(2);
    std::cout << "\n=== Monthly DCA of " << monthly_amount << " SAR into Maaden (1211.SR) [C++] ===\n";
    std::cout << "Period: " << rows.front().date << " to " << rows.back().date
              << " (" << months << " contributions)\n";
    std::cout << "Total invested: " << strat.total_invested << " SAR\n\n";

    std::cout << "--- Strategy (SMA crossover + RSI + trailing stop) ---\n";
    std::cout << "Final value: " << strat.final_value << " SAR\n";
    std::cout << "Profit:      " << strat.profit << " SAR (" << strat.return_pct << "%)\n";
    std::cout << "Trades: " << strat.num_trades << ", Win rate: " << strat.win_rate_pct << "%\n\n";

    std::cout << "--- Buy & hold (invest every contribution immediately) ---\n";
    std::cout << "Final value: " << bh.final_value << " SAR\n";
    std::cout << "Profit:      " << bh.profit << " SAR (" << bh.return_pct << "%)\n";
    std::cout << "Shares held: " << static_cast<long long>(bh.shares) << "\n";

    return 0;
}
